metricproviders/plugin/client/client.go

```diff
@@ -0,0 +1,89 @@
+package client
+
+import (
+	"fmt"
+	"os/exec"
+
+	"github.com/argoproj/argo-rollouts/metricproviders/plugin/rpc"
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	"github.com/argoproj/argo-rollouts/utils/plugin"
+	goPlugin "github.com/hashicorp/go-plugin"
+)
+
+type singletonMetricPlugin struct {
+	pluginClient map[string]*goPlugin.Client
+	plugin       map[string]rpc.MetricsPlugin
+}
+
+var singletonPluginClient *singletonMetricPlugin
+
+// GetMetricPlugin returns a singleton plugin client for the given metric plugin. Calling this multi
+// returns the same plugin client for the same plugin name.
+func GetMetricPlugin(metric v1alpha1.Metric) (rpc.MetricsPlugin, error) {
+	if singletonPluginClient == nil {
+		singletonPluginClient = &singletonMetricPlugin{
+			pluginClient: make(map[string]*goPlugin.Client),
+			plugin:       make(map[string]rpc.MetricsPlugin),
+		}
+	}
+	plugin, err := singletonPluginClient.startPluginSystem(metric)
+	if err != nil {
+		return nil, fmt.Errorf("unable to start plugin system: %w", err)
+	}
+	return plugin, nil
+}
+
+func (m *singletonMetricPlugin) startPluginSystem(metric v1alpha1.Metric) (rpc.MetricsPlugin, error) {
```

Comment:
It seems that all this function is doing is registering a new metric plugin? If so, can we rename the method to `register` maybe? `startPluginSystem` gives the impression that many plugins will be initialized.

Reply:
This starts the child process of the plugin if it is not running already and returns the existing client if it is already running. It can be started multiple times if there are multiple plugins configured once for each plugin name configured.